base/iterators.jl

```diff
@@ -1202,7 +1203,19 @@ julia> [(x,y) for x in 0:1 for y in 'a':'c']  # collects generators involving It
 flatten(itr) = Flatten(itr)
 
 eltype(::Type{Flatten{I}}) where {I} = eltype(eltype(I))
-eltype(::Type{Flatten{I}}) where {I<:Union{Tuple,NamedTuple}} = promote_typejoin(map(eltype, fieldtypes(I))...)
+
+# For tuples, we statically know the element type of each index, so we can compute
+# this at compile time.
+function eltype(::Type{Flatten{I}}) where {I<:Union{Tuple,NamedTuple}}
+    @_foldable_meta
```

Comment:
```suggestion
```

I think it's not permissible to assume :foldable, because the loop body calls `eltype` on user types, so it calls arbitrary user-defined code.

Reply:
How about
```julia
eltype(::Type{Flatten{I}}) where {I<:Union{Tuple,NamedTuple}} = 
    afoldl((T, i) -> Base.promote_typejoin(T, Base.eltype(i)), Union{}, fieldtypes(I)...)
```